core/src/main/java/io/grpc/inprocess/InProcessTransport.java

```diff
@@ -557,10 +564,19 @@ public void start(ClientStreamListener listener) {
         synchronized (InProcessTransport.this) {
           ServerStreamListener serverStreamListener = serverTransportListener.streamCreated(
```

Comment:
I'd much rather solve the race by splitting initialization and start. Basically allow `clientStream.setListener()` to be called and then have a separate start() which calls into the application.